web/components/roomConfig/GitlabRepoConfig.tsx

```diff
@@ -0,0 +1,167 @@
+import { h, FunctionComponent, createRef } from "preact";
+import { useState, useCallback, useEffect } from "preact/hooks";
+import BridgeAPI from "../../BridgeAPI";
+import { ConnectionConfigurationProps, RoomConfig } from "./RoomConfig";
+import { GitLabRepoConnectionState, GitLabRepoResponseItem, GitLabTargetFilter, GitLabRepoConnectionTarget, GitLabRepoConnectionProjectTarget, GitLabRepoConnectionInstanceTarget } from "../../../src/Connections/GitlabRepo";
+import InputField from "../InputField";
+import ButtonSet from "../ButtonSet";
+import { Button } from "../Button";
+
+const EventType = "uk.half-shot.matrix-hookshot.gitlab.repository";
+
+const ConnectionSearch: FunctionComponent<{api: BridgeAPI, onPicked: (state: GitLabRepoConnectionState) => void}> = ({api, onPicked}) => {
+    const [filter, setFilter] = useState<GitLabTargetFilter>({});
+    const [results, setResults] = useState<GitLabRepoConnectionProjectTarget[]|null>(null);
+    const [instances, setInstances] = useState<GitLabRepoConnectionInstanceTarget[]|null>(null);
+    const [debounceTimer, setDebounceTimer] = useState<number>(null);
+    const [currentProjectPath, setCurrentProjectPath] = useState<string|null>(null);
+
+    const searchFn = useCallback(() => {
+        api.getConnectionTargets<GitLabRepoConnectionTarget>(EventType, filter).then(res => {
+            if (!filter.instance) {
+                setInstances(res);
+                if (res[0]) {
+                    setFilter({instance: res[0].name});
+                }
+            } else {
+                setResults(res as GitLabRepoConnectionProjectTarget[]);
+            }
+        });
+    }, [api, filter]);
+
+    const updateSearchFn = useCallback((evt: InputEvent) => {
+        setFilter({...filter, search: (evt.target as HTMLInputElement).value });
+    }, [filter, setFilter]);
+
+    useEffect(() => {
+        if (debounceTimer) {
+            clearTimeout(debounceTimer);
+        }
+        // Browser types
+        setDebounceTimer(setTimeout(searchFn, 500) as unknown as number);
+        // Things break if we depend on the thing we are clearing.
+        // eslint-disable-next-line react-hooks/exhaustive-deps
```

Comment:
What kind of dark React magic is being performed here??
If it works, I'll pretend not to care, but this is how you'd usually clean up Effects:

You `return` a function that undoes your changes.
https://gitlab.com/jaller94/matrix-wrench/-/blob/9f045b8926802fde53c7097c8b26bb58d632b88d/src/index.js#L392-395

Reply:
Good shout on the return, have done that.

On the dark magic, I'm basically trying to write a debounce fn that clears the existing timeout when this effect is fired. The problem with including the `debounceTimer` value is that when we set the value of `debounceTimer`, the effect is presumably recalled leading to a loop?

If I am wrong on that one, we can probably drop this ignore line.